src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -0,0 +1,287 @@
+  }
+
+  def polymorphicSkat[T <: Vector[Double]](vds: VariantDataset,
+    keyName: String,
+    variantKeys: String,
+    singleKey: Boolean,
+    weightExpr: Option[String],
+    yExpr: String,
+    covExpr: Array[String],
+    getGenotypes: (Iterable[Genotype], Int) => T,
+    zero: SkatAgg[T],
+    resultOp:(SkatAgg[T], Double) => SkatStat): KeyTable =
+  {
+
+    val (y, cov, completeSamples) = RegressionUtils.getPhenoCovCompleteSamples(vds, yExpr, covExpr)
+
+    val n = y.size
+    val k = cov.cols
+    val d = n - k
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+
+    // fit null model
+    val qr.QR(q, r) = qr.reduced.impl_reduced_DM_Double(cov)
+    val beta = r \ (q.t * y)
+    val res = y - cov * beta
+    val sigmaSq = (res dot res) / d
+
+    val filteredVds = weightExpr match {
+      case None =>  vds.filterSamplesList(completeSamples.toSet)
+                       .annotateVariantsExpr("va.AF = gs.callStats(g=> v).AF")
+                       .annotateVariantsExpr("va.weight = let af = if (va.AF[0] <= va.AF[1]) va.AF[0] else va.AF[1] in dbeta(af,1.0,25.0)**2")
+      case _ => vds.filterSamplesList(completeSamples.toSet)
+    }
+
+
+    val (keysType, keysQuerier) = filteredVds.queryVA(variantKeys)
+    val (weightType, weightQuerier) =  weightExpr match {
+      case None => filteredVds.queryVA("va.weight")
+      case _    => filteredVds.queryVA(weightExpr.get)
+    }
+
+    val typedWeightQuerier = weightType match {
+      case TFloat64 => weightQuerier.asInstanceOf[Annotation => Double]
+      case TFloat32 => weightQuerier.asInstanceOf[Annotation => Double]
+      case TInt64 => weightQuerier.asInstanceOf[Annotation => Double]
+      case TInt32 => weightQuerier.asInstanceOf[Annotation => Double]
+      case _ => fatal("Weight must evaluate to numeric type")
+    }
+
+    val sc = filteredVds.sparkContext
+    val resBc = sc.broadcast(res)
+
+    def variantPreProcess(gs: Iterable[Genotype], w: Double):SkatTuple[T] = {
+      val sqrtw = math.sqrt(w.asInstanceOf[Double])
+      val wx: T = (getGenotypes(gs, n) * sqrtw).asInstanceOf[T]
+      val Sj = resBc.value dot wx
+      SkatTuple(Sj * Sj, wx, q.t * wx)
+    }
+
+    val (keyType, keyIterator): (Any, Any => Iterator[Any]) = if (singleKey) {
+      (keysType, (key: Any) => Iterator.single(key))
+    } else {
+      val keyType = keysType match {
+        case TArray(e) => e
+        case TSet(e) => e
+        case _ => fatal(s"With single_key=False, variant keys must be of type Set[T] or Array[T], got $keysType")
+      }
+      (keyType, (keys: Any) => keys.asInstanceOf[Iterable[Any]].iterator)
+    }
+
+    val keyedRdd = filteredVds.rdd.flatMap { case (v, (va, gs)) =>
+      (Option(keysQuerier(va)), Option(typedWeightQuerier(va))) match {
+        case (Some(key), Some(w)) =>
+          keyIterator(key).map((_, variantPreProcess(gs, w.asInstanceOf[Double])))
```

Comment:
I could be wrong, but isn't this going to "preprocess" the same row of genotypes multiple times if a variant is in multiple genes? Shouldn't it be processed once and then duplicated afterwards?